src/main/java/com/intuit/graphql/orchestrator/batch/QueryOptimizer.java

```diff
@@ -6,99 +6,68 @@
 import graphql.language.SelectionSet;
 
 import java.util.ArrayList;
-import java.util.Collection;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.List;
-import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.Function;
-import java.util.function.Predicate;
-import java.util.stream.Collectors;
+import java.util.Map;
+import org.apache.commons.lang3.tuple.ImmutablePair;
 
 public class QueryOptimizer {
 
-    private final Operation operationType;
-    private final SelectionSet filteredSelection;
+  private final Operation operationType;
 
-    public QueryOptimizer(Operation operationType, SelectionSet filteredSelection) {
-        this.filteredSelection = filteredSelection;
-        this.operationType = operationType;
-    }
-
-    public static <T> Predicate<T> distinctByFieldName(Function<? super T, ?> keyExtractor) {
-        Set<Object> seen = ConcurrentHashMap.newKeySet();
-        return t -> seen.add(keyExtractor.apply(t));
-    }
+  public QueryOptimizer(Operation operationType) {
+    this.operationType = operationType;
+  }
 
-    public static void createSelectionSetTree(SelectionSet selectionSet, HashMap<String, Set<Field>> selectionSetMap) {
-        selectionSet.getSelections().forEach(field -> {
-            Field f = (Field) field;
-            if (f.getSelectionSet() != null) {
-                if (selectionSetMap.containsKey(f.getName()))
-                    selectionSetMap.get(f.getName()).addAll(getSelectionSetFields(f));
-                else
-                    selectionSetMap.put(f.getName(), new HashSet<>(getSelectionSetFields(f)));
-                createSelectionSetTree(f.getSelectionSet(), selectionSetMap);
-            }
-        });
-    }
+  public static ImmutablePair<List<Selection>, Map<String, List<Field>>> groupSelections(List<Selection> selections) {
+    List<Selection> distinctSelections = new ArrayList<>();
+    Map<String, List<Field>> fieldMap = new HashMap<>();
```

Comment:
Should we use a [LinkedHashMap](http://docs.oracle.com/javase/7/docs/api/java/util/LinkedHashMap.html) to maintain the order? It might not be necessary, just a thought.